docs(pulsar): add CloudEvents schema wrapping, rawschema option (#5092)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-pubsub/setup-pulsar.md

| batchingMaxSize | N | batchingMaxSize sets the maximum number of bytes permitted in a batch. If set to a value greater than 1, messages will be queued until this threshold is reached or batchingMaxMessages (see above) has been reached or the batch interval has elapsed. Default: `"128KB"` | `"131072"`|
| <topic-name>.jsonschema          | N  | Enforces JSON schema validation for the configured topic. When CloudEvents wrapping is enabled (the default), the schema registered with the Pulsar Schema Registry is a CloudEvents envelope JSON schema containing the provided schema as the `data` field. See [Publishing & subscribing messages with Cloudevents]({{% ref pubsub-cloudevents.md %}})| |
| <topic-name>.avroschema          | N  | Enforces Avro schema validation for the configured topic. When CloudEvents wrapping is enabled (the default), the schema registered with the Pulsar Schema Registry is a CloudEvents envelope Avro schema containing the provided schema as the `data` field. See [Publishing & subscribing messages with Cloudevents]({{% ref pubsub-cloudevents.md %}}) | |

> | <topic-name>.rawschema          | N  | When set to `"true"`, registers the inner schema (Avro or JSON) directly with the Pulsar Schema Registry instead of wrapping it in a CloudEvents envelope. Use this for topics that exclusively receive raw payloads. Callers must also set the publisher request metadata `rawPayload=true`. See [Publishing & subscribing messages without CloudEvents]({{% ref pubsub-raw.md %}}). Default: `"false"` | `"true"`, `"false"` |

Reviewer: ```suggestion
| <topic-name>.rawschema          | N  | When set to `"true"`, registers the raw message schema (Avro or JSON) directly with the Pulsar Schema Registry instead of wrapping it in a CloudEvents envelope. Use this for topics that exclusively receive raw payloads. Callers must also set the publisher request metadata `rawPayload=true`. See [Publishing & subscribing messages without CloudEvents]({{% ref pubsub-raw.md %}}). Default: `"false"` | `"true"`, `"false"` |
```

    value: "openid,profile,email"
```

> ### Schema validation and CloudEvents envelope

Reviewer: ```suggestion
### Use JSON/Avro schema validation
```

> Dapr allows you to perform schema validation on Avro and JSON message schema using the `<topic-name>.avroschema` and `<topic-name>.jsonschema` metadata properties. When using this feature with CloudEvents enabled (the [default]({{% ref pubsub-cloudevents.md %}})), Dapr automatically wraps your schema inside a CloudEvents envelope before registering it with the Pulsar Schema Registry. For Avro schemas, the envelope follows the [CloudEvents Avro format spec](https://github.com/cloudevents/spec/blob/main/cloudevents/bindings/avro-format.md). For JSON schemas, the envelope follows the [CloudEvents JSON format spec](https://github.com/cloudevents/spec/blob/main/cloudevents/formats/json-format.md). This ensures the registered Pulsar schema matches the actual wire format of published messages.

Reviewer: ```suggestion
Dapr allows you to perform schema validation on Avro and JSON message schema using the `<topic-name>.avroschema` and `<topic-name>.jsonschema` metadata properties. When using this feature with CloudEvents enabled (the [default]({{% ref pubsub-cloudevents.md %}})), Dapr automatically wraps your schema inside a CloudEvents envelope before registering it with the Pulsar Schema Registry. When using this feature with [raw messages]({{% ref pubsub-raw.md %}}), Dapr registers your raw message schema with the Pulsar Schema Registry.

For Avro schemas, the envelope follows the [CloudEvents Avro format spec](https://github.com/cloudevents/spec/blob/main/cloudevents/bindings/avro-format.md). For JSON schemas, the envelope follows the [CloudEvents JSON format spec](https://github.com/cloudevents/spec/blob/main/cloudevents/formats/json-format.md). This ensures the registered Pulsar schema matches the actual wire format of published messages.
```

The following table summarizes the behavior (applies to both `.avroschema` and `.jsonschema`):

| | CloudEvents wrapper (default) | Raw messages |
|---|---|---|
| **No schema validation configured** | Dapr wraps message in a CloudEvents envelope (JSON). Message sent as raw bytes as the `data` field. No schema registered with Pulsar. | Dapr sends the message as raw bytes without schema validation or CloudEvents envelope. |
| **Schema validation with `.rawschema=false`** | Dapr wraps message in CloudEvents envelope. Publisher registers CloudEvents envelope schema (either JSON or Avro) and validates against CloudEvents codec. | Rejected with error. Publisher schema is a CloudEvents envelope and so raw payload fails validation. |
| **Schema validation with `.rawschema=true`** | Rejected with error. Publisher registers raw data with Pulsar but Dapr wraps in CloudEvents envelope and so fails validation. | Publisher registers raw data with Pulsar and validates against inner codec. |

{{% alert title="Important" color="warning" %}}

> Because Dapr wraps messages in a CloudEvents envelope by default, never use `.rawschema=true` on a topic that will receive default (non-`rawPayload`) publishes. The `.rawschema=true` flag must only be used together with `rawPayload=true` publishes on a dedicated topic, so that the registered inner schema always matches the actual wire format.

Reviewer: @javier-aliaga this comment wasnt copied over from then other PR

Reviewer: Javi i dont think that this was copied over?? It should be

Because Dapr wraps messages in a CloudEvents envelope by default, never use `<topic-name>.rawschema=true` on a topic that will receive CloudEvent wrapped messages. The `<topic-name>.rawschema=true` metadata property must only be used together with the publisher setting`rawPayload=true`, on a dedicated topic, so the registered inner data schema always matches the actual wire format. Read [Publishing & subscribing messages without CloudEvents]({{% ref pubsub-raw.md %}}) for how to use `rawPayload=true`.

{{% /alert %}}

Since Pulsar enforces a single schema per topic, `rawPayload=true` messages cannot be sent to a topic using a CloudEvents envelope schema. To handle both CE-wrapped and raw payloads, use separate topics:

> ```yaml

Reviewer: This is too confusing of an example for users to understand. Can we split this into to two Component examples under two headers:

### Schema validation with CloudEvents

### Schema validation with raw messages

apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: pulsar-pubsub
spec:
  type: pubsub.pulsar
  version: v1
  metadata:
  - name: host
    value: "localhost:6650"
  # CE-wrapped topic (default behavior) — Avro example
  - name: orders.avroschema
    value: '{"type":"record","name":"Order","fields":[{"name":"ID","type":"int"},{"name":"Name","type":"string"}]}'

>   # Raw-payload-only topic (skips CE wrapping) — Avro example

Reviewer: ```suggestion
  # Raw-payload-only topic (skips CloudEvent wrapping) — Avro example
```

Reviewer: In general we shoudn't shorten CloudEvent to CE

    value: "openid,profile,email"
```

### Use JSON/Avro schema validation

Dapr allows you to perform schema validation on Avro and JSON message schema using the `<topic-name>.avroschema` and `<topic-name>.jsonschema` metadata properties. When using this feature with CloudEvents enabled (the [default]({{% ref pubsub-cloudevents.md %}})), Dapr automatically wraps your schema inside a CloudEvents envelope before registering it with the Pulsar Schema Registry. When using this feature with [raw messages]({{% ref pubsub-raw.md %}}), Dapr registers your raw message schema with the Pulsar Schema Registry.

For Avro schemas, the envelope follows the [CloudEvents Avro format spec](https://github.com/cloudevents/spec/blob/main/cloudevents/bindings/avro-format.md). For JSON schemas, the envelope follows the [CloudEvents JSON format spec](https://github.com/cloudevents/spec/blob/main/cloudevents/formats/json-format.md). This ensures the registered Pulsar schema matches the actual wire format of published messages.

The following table summarizes the behavior (applies to both `.avroschema` and `.jsonschema`):

| | CloudEvents wrapper (default) | Raw messages |
|---|---|---|

> | **No schema validation configured** | Dapr wraps message in a CloudEvents envelope (JSON). Message sent as raw bytes as the `data` field. No schema registered with Pulsar. | Dapr sends the message as raw bytes without schema validation or CloudEvents envelope. |

Reviewer: ```suggestion
| **No schema validation configured** | Dapr wraps the message in a CloudEvents envelope (JSON). Message sent as raw bytes as the `data` field. No schema registered with Pulsar. | Dapr sends the message as raw bytes without schema validation or CloudEvents envelope. |
```

> | **Schema validation with `.rawschema=false`** | Dapr wraps message in CloudEvents envelope. Publisher registers CloudEvents envelope schema (either JSON or Avro) and validates against CloudEvents codec. | Rejected with error. Publisher schema is a CloudEvents envelope and so raw payload fails validation. |

Reviewer: ```suggestion
| **Schema validation with `<topic-name>.rawschema=false`** | Dapr wraps message in CloudEvents envelope. Publisher registers CloudEvents envelope schema (either JSON or Avro) and validates against CloudEvents codec. | Rejected with error. Publisher schema is a CloudEvents envelope and so raw payload fails validation. |
```

> | **Schema validation with `.rawschema=true`** | Rejected with error. Publisher registers raw data with Pulsar but Dapr wraps in CloudEvents envelope and so fails validation. | Publisher registers raw data with Pulsar and validates against inner codec. |

Reviewer: ```suggestion
| **Schema validation with `<topic-name>.rawschema=true`** | Rejected with error. Publisher registers raw payload with Pulsar but Dapr wraps it in a CloudEvents envelope and so fails validation. | Publisher registers raw data with Pulsar and validates it against inner codec. |
```

{{% alert title="Important" color="warning" %}}
Because Dapr wraps messages in a CloudEvents envelope by default, never use `.rawschema=true` on a topic that will receive default (non-`rawPayload`) publishes. The `.rawschema=true` flag must only be used together with `rawPayload=true` publishes on a dedicated topic, so that the registered inner schema always matches the actual wire format.
{{% /alert %}}

Since Pulsar enforces a single schema per topic, raw messages cannot be sent to a topic using a CloudEvents envelope schema. To handle both CloudEvent-wrapped and raw payloads, use separate topics.

#### Schema validation with CloudEvents

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: pulsar-pubsub
spec:
  type: pubsub.pulsar
  version: v1
  metadata:
  - name: host
    value: "localhost:6650"
  # CloudEvent-wrapped topic (default behavior) — Avro example
  - name: orders.avroschema
    value: '{"type":"record","name":"Order","fields":[{"name":"ID","type":"int"},{"name":"Name","type":"string"}]}'
```

> 

Reviewer: Publishing with `rawPayload=true` to a CloudEvent-wrapped topic returns the error:
> `rawPayload=true is not compatible with schema topics using CloudEvents envelope; use a separate topic for raw payloads`

Reviewer: This section should be here

#### Schema validation with raw messages

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: pulsar-pubsub
spec:
  type: pubsub.pulsar
  version: v1
  metadata:
  - name: host
    value: "localhost:6650"
  # Raw-payload-only topic (skips CloudEvent wrapping) — Avro example
  - name: orders-raw.avroschema
    value: '{"type":"record","name":"Order","fields":[{"name":"ID","type":"int"},{"name":"Name","type":"string"}]}'
  - name: orders-raw.rawschema
    value: "true"
  # Raw-payload-only topic — JSON schema example
  - name: events-raw.jsonschema
    value: '{"type":"object","properties":{"id":{"type":"integer"},"name":{"type":"string"}},"required":["id","name"]}'
  - name: events-raw.rawschema
    value: "true"
```

> Publishing with `rawPayload=true` to a CloudEvent-wrapped topic returns the error:

Reviewer: ```suggestion
```

> > `rawPayload=true is not compatible with schema topics using CloudEvents envelope; use a separate topic for raw payloads`

Reviewer: ```suggestion
```

The following table summarizes the behavior (applies to both `.avroschema` and `.jsonschema`):

| | CloudEvents wrapper (default) | Raw messages |
|---|---|---|
| **No schema validation configured** | Dapr wraps the message in a CloudEvents envelope (JSON). Message sent as raw bytes as the `data` field. No schema registered with Pulsar. | Dapr sends the message as raw bytes without schema validation or CloudEvents envelope. |
| **Schema validation with `<topic-name>.rawschema=false`** | Dapr wraps message in CloudEvents envelope. Publisher registers CloudEvents envelope schema (either JSON or Avro) and validates against CloudEvents codec. | Rejected with error. Publisher schema is a CloudEvents envelope and so raw payload fails validation. |
| **Schema validation with `<topic-name>.rawschema=true`** | Rejected with error. Publisher registers raw payload with Pulsar but Dapr wraps it in a CloudEvents envelope and so fails validation. | Publisher registers raw data with Pulsar and validates it against inner codec. |

{{% alert title="Important" color="warning" %}}

> Because Dapr wraps messages in a CloudEvents envelope by default, never use `.rawschema=true` on a topic that will receive default (non-`rawPayload`) publishes. The `.rawschema=true` flag must only be used together with `rawPayload=true` publishes on a dedicated topic, so that the registered inner schema always matches the actual wire format.

Reviewer: ```suggestion
Because Dapr wraps messages in a CloudEvents envelope by default, never use `<topic-name>.rawschema=true` on a topic that will receive CloudEvent wrapped messages. The `<topic-name>.rawschema=true` metadata property must only be used together with the publisher setting`rawPayload=true`, on a dedicated topic, so the registered inner data schema always matches the actual wire format. Read [Publishing & subscribing messages without CloudEvents]({{% ref pubsub-raw.md %}}) for how to use `rawPayload=true`.
```
